Public/Src/Cache/MemoizationStore/Distributed/Stores/RedisMemoizationDatabase.cs

```diff
@@ -0,0 +1,192 @@
+                    return new Result<bool>(result);
+                });
+        }
+
+        /// <inheritdoc />
+        public override Task<IEnumerable<StructResult<StrongFingerprint>>> EnumerateStrongFingerprintsAsync(OperationContext context)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override async Task<Result<(ContentHashListWithDeterminism contentHashListInfo, string replacementToken)>> GetContentHashListAsync(OperationContext context, StrongFingerprint strongFingerprint)
+        {
+            var key = GetKey(strongFingerprint.WeakFingerprint);
+
+            (byte[] metadataBytes, string replacementToken) = await _redis.ExecuteBatchAsync(context,
+                async batch =>
+                {
+                    byte[] selectorBytes = SerializeSelector(strongFingerprint.Selector, isHash: false);
+                    var metadataBytesTask = batch.AddOperation(key, b => b.HashGetAsync(key, selectorBytes));
+
+                    byte[] replacementTokenFieldName = SerializeSelector(strongFingerprint.Selector, isHash: true);
+                    var replacementTokenTask = batch.AddOperation(key, b => b.HashGetAsync(key, replacementTokenFieldName));
+
+                    return ((byte[])await metadataBytesTask, (string)await replacementTokenTask);
+                },
+                RedisOperation.GetContentHashList);
+
+            if (metadataBytes == null)
+            {
+                return new Result<(ContentHashListWithDeterminism contentHashListInfo, string replacementToken)>((default, string.Empty));
+            }
+
+            var metadata = DeserializeMetadataEntry(metadataBytes);
+
+            // Update the time, only if no one else has changed it in the mean time. We don't
+            // really care if this succeeds or not, because if it doesn't it only means someone
+            // else changed the stored value before this operation but after it was read.
+            CompareExchange(context, strongFingerprint, replacementToken, metadata.ContentHashListWithDeterminism, metadata.ContentHashListWithDeterminism, replacementToken).FireAndForget(context);
+
+            return new Result<(ContentHashListWithDeterminism, string)>((metadata.ContentHashListWithDeterminism, replacementToken));
+        }
+
+        /// <inheritdoc />
+        public override async Task<Result<LevelSelectors>> GetLevelSelectorsAsync(OperationContext context, Fingerprint weakFingerprint, int level)
+        {
+            var weakFingerprintKey = GetKey(weakFingerprint);
+
+            var redisValues = await _redis.GetHashKeysAsync(context, weakFingerprintKey, context.Token);
+
+            var result = new List<Selector>(redisValues.Length / 2);

```

Comment:
Question: Why is it sized to half the results?

Reply:
Because of the schema that we chose. To avoid comparing byte[], we upload a hash of the contenthashlist, to be able to only compare strings.

You can imagine the schema to be: WeakFingerprint -> Dictionary<Selector, (ContentHashList:byte[], Hash:string)>, with the difference that, since Redis does not have the concept of value tuple, we duplicate each Selector, prefixed with either a 1 or a 0, depending if we're storing the ContentHasList or the hash

Reply:
Rather than a comment in the PR, add this to the a comment on the method or class. Also, its a guid rather than a hash of the content hash list now.

---
In reply to: [320996591](https://github.com/microsoft/BuildXL/pull/832#discussion_r320996591) [](ancestors = 320996591)